### ISSUE TITLE
Plugin Development - Various Issues relating to loading the assemblies

### ISSUE THREAD

**Issue by piercy** [antonpup/Aurora#1010]
Hey,

So ive been working on a plugin to communicate with an outside API and do some stuff on my keyboard.  For the most part its working but i'm having several issues that are to do with the plugin architecture.  This is not a criticism, but im wondering if im the first person to really use the Plugin stuff, as i feel like these issues cant be worked around and would likely have come up for other people.

So, Aurora starts, with the plugin loaded but no layers yet created.  I go in, create my layer(s), test them, all works and i exit. This part is fine.

Next, i start up Aurora. This time round, it loads the layer from the profile config and throws an error (its caught but it still throws it).  "Newtonsoft Json - Could not load file or assembly Plugin-pluginName".  
 As its thrown, the layer doesnt load and it just disappears.  From my understanding, this is because it doesn't know where the plugin assembly exists.  It doesn't know about the Plugins folder and even if it did, i dont think NewtonsoftJson is equipped to load it from another folder.  I know .NET tends to have issues loading assemblies in other folders, so i moved on and tried something different.

Next, i put my plugin assembly in the root folder as well as the plugins folder.  This appeared to work at first. However, i quickly(took all night tbh..) realised that if Aurora starts up with layers already added, when you try to add another you get an error about the URI not existing for the XAML file.  Investigating this further leads me to believe that Aurora was loading the XAML file from the root area (and this means URI looks something like "component;Aurora/something/something.xaml") and when you try to load the XAML file in another way the URI does not start with Aurora (because its now from the plugin so its more like "component;Plugin-pluginName/something/something.xaml") and thus it fails to find it.

So, I cant put it in just the plugins folder and i cant put it in both folders.  So next i tried to stick it in just the root folder. This had some issues ... i had to change the const in the plugin manager so that it didn't look in the Plugins directory and i also had to filter it so that it only looks for "plugin-*.dll" otherwise it was loading the entire root directories assemblies.

This works, the layers save and load, the plugins are loaded in and out correctly all is working.

Now, i don't know if this is an acceptable solution.  Its much cleaner having everything in the plugins directory. Also it does mean, I am having to make modifications to the plugin manager which could potentially break it for other people who have made plugins (if there are any.. i suspect not as I would expect this issue to come up pretty quickly).

Also, you can do all the steps i provided above with the Plugin-Example.  So if you'd like to test you can try it with that plugin.

I hope this all makes sense.  I'm looking for advice on how to proceed really. 

Cheers,
Piercy

**Comment by piercy** [comment]
To give some more context to this, tonight i basically finished off the plugin other than the above modifications i had to make.  So i now have a plugin that can receive any notification from a mobile phone and trigger some layer effects.  I will add more lighting effects, at the moment it only uses the breathing effect.

All keys and colours are configurable but as examples here's how i'm using it currently.

If i receive a phone call, my whole keyboard breathes white.
Facebook, Snapchat, Whatsapp and Email notifications trigger my F1-F12 keys to breath their respective apps colour (Blue, Yellow, Green and Red).

Although i pre-configure these notifications, the software can work with any apps notifications and should be both IOS and Android compatible although i don't have an IOS device to test.

**Comment by simon-wh** [comment]
I had a quick look at some of the changes you've made in your fork. I would like to keep the plugins in a separate folder, rather than them being prefixed with "Plugin-", it's just better for the end-user as it is easier for them to keep track of what plugins they have. The state of the plugin system (the way it is in the current dev branch) is nowhere near the finished setup, it was something I had worked on before and hadn't fully finished as I was focusing on other things, hence the fact it isn't mentioned anywhere. I plan to make plugins be able to implement fully functioning Applications and Devices (exactly the same way they exist in the application), similar to how it is currently setup for the plugins to create additional layers. If you have any ideas for improvements that can be made to the system, feel free to post them, or if you wanted, you could implement some, but that's up to you.

**Comment by piercy** [comment]
Thanks for taking a look.  I agree, it is much messier would much prefer it to be in the Plugins folder.   Also,  keeping the code line separate for plugins would be great as well, allows for different code stlye and less tech support needed in the main application. 

I spoke to a colleague at work about the issue i was experiencing. He sent me some code that may potentially work for telling Json how to handle the plugin assemblies.  I look into this and test it with the example plugin (to make sure that im not just fixing in suitable for myself).  If i manage to do it ill submit a pull request.

One other thing i did come across that's worth keeping in mind when you revisit plugins.  Currently, all the interfaces, classes and other code needed for plugins exists inside the main project-aurora project.  This means you need to reference the full project to be able to develop a plugin.  Ideally, some of the bits need pulling out to a separate assembly so other projects can reference the assembly.  Its not important for now but something to bear in mind if/when you swing back round.


Thanks again for taking a look.

**Comment by simon-wh** [comment]
Awesome!

I think I might try and separate the main parts of the application into:
UI,
Core Functionality,
And use plugins for Applications, Layers and Devices

By moving that stuff into plugins it'll make it easier to ensure that the plugin system works well.

However, that's just an idea, let me know what you think of it. I reckon I'd likely change this up when I switch to AvaloniaUI for the ui to start attempting to make Aurora work cross-platform.

**Comment by piercy** [comment]
Yeah, that all sounds good really.  If you were to look at the existing game handlers, and others bits and bobs.  You can clearly see how you can apply the same thing in the plugins to achieve the same result.  I mean, most of my code i wrote was from looking around the project... looking at how X stores settings, how X triggers certain lighting effects. After that i just communicate with an API so i know when notifications are received on my phone.

I think regarding new games and applications plugins are an easy place to start.  You basically grab some of the existing Handlers and UI's that exist. Whack them in a plugin and hook up to the new games API somehow. Whether that's using existing wrappers or specifically coding to a games API.  After that your mostly there.   I think my plugin took me about 3 or 4 days but i was learning as i went.  I think if you are familiar with it, That time could drastically come down meaning new games can be added very easily (provided some form of API support from the game).


So i'm pretty sure a lot of the existing Applications, Layers and Devices will work well.  The tricky part is the cohesion between them.  You don't want people repeating code because they also want say "a rainbow" layer or something silly.  You'll end up with 30 different rainbow effects.  I mean, people could just reference the base plugins but it essentially becomes working out what should be core and what should be plugin.

its almost more a conceptual issue than it is an actual code issue.  But really, for the most part i think plugins fit well with this. You can see in the current project how lots of different things are designed very similar and just slot in. 

Side note, I haven't yet tried the code a colleague sent me.  GDPR is kicking my arse at work so i havent really felt like spending time doing more code but i will definitely give it a shot to see if i can solve the saving and loading issue.  That said, i'm happily using the rubbish build i wrote for now.  Havent missed a phone call since doing it so its at least working.  Ill make sure to post back if i find anything .

**Comment by simon-wh** [comment]
Yeah, the current way the Applications and Layer Handlers exist make it quite easy to make new ones, especially if they use an existing wrapper, in that case it's just a matter of copying everything, which now, isn't much and changing references to the specific game and adding the appropriate image.

Right now the Application Profiles by default will create particular layers which somewhat throws a spanner in the idea of separating all that stuff into plugins as they would still need to reference eachother. However, I reckon for the core stuff I'd likely make it so that they run through the Plugin system but are required, i.e. the application won't run without them and they can't be disabled, but they will still be executed through the system to keep continuity. With that, it should be safe for those assemblies to be referenced by each other or other plugins so they can add particular layers they want by default. I'll likely try and keep them separated as there really should only be referenced from the Applications to the Layer Handlers, never the other way around, unless they are ones that are only to be used with that application. I reckon that everything in the current project that is going to be set up to be added as plugins should be moved into plugin form.

In theory when it comes to user-made plugins, if someone was to make a layer handler pack for example, someone who wants to use some of them as default layers for an Application they're adding support for should be able to reference the assembly for the pack and use it, and require that the pack is installed for the application to work.

There's no pressure on you to work on this, any help is appreciated, if you don't do it I'll get to it hopefully soon (but considering how I'm getting to things a lot later than I intend it might not be for a while)

**Comment by piercy** [comment]
im happy to work on it.. Id like to get my build into using the plugins properly.  As im mostly there and have sample code that may work least i can do is try it.
